### ISSUE TITLE
JoyPro crash when importing bindings

### ISSUE THREAD

**Issue by Chaussette23** [Holdi601/JoystickProfiler#60]
Hello,

I use this wonderful tool since some years now, and I have a problem which looks quite new : I updated to latest release (91) by overwritting my JoyPro folder by the cone contained in the archive.

I use DCS openbeta and I'm up to date

If I launch JoyPro and check the log file, I get a lot of errors in clean profiles :
```
[22/09/23 17:21:28]: 	Load default DCS: E:\Utilities\JoyPro\CleanProfile\DCS\A-10C II.cf
[22/09/23 17:21:28]: 	Load default DCS: E:\Utilities\JoyPro\CleanProfile\DCS\A-10C II_easy.cf
[22/09/23 17:21:28]: 	System.FormatException: Input string was not in a correct format.
   at System.Number.ParseDouble(String value, NumberStyles options, NumberFormatInfo numfmt)
   at System.Convert.ToDouble(String value, IFormatProvider provider)
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 112
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 105
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 105
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 105
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 105
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 105
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 105
   at JoyPro.DCSLuaInput.AnalyzeRawLuaInput(String content, DCSExportPlane refMod) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\DataStructures\DCS\DCSLuaInput.cs:line 333
   at JoyPro.DCSIOLogic.LoadCleanLuasDCS() in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\DataStructures\DCS\DCSIOLogic.cs:line 258
[22/09/23 17:21:28]: 	Input string was not in a correct format.
[22/09/23 17:21:28]: 	mscorlib
[22/09/23 17:21:28]: 	   at System.Number.ParseDouble(String value, NumberStyles options, NumberFormatInfo numfmt)
   at System.Convert.ToDouble(String value, IFormatProvider provider)
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 112
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 105
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 105
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 105
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 105
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 105
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 105
   at JoyPro.DCSLuaInput.AnalyzeRawLuaInput(String content, DCSExportPlane refMod) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\DataStructures\DCS\DCSLuaInput.cs:line 333
   at JoyPro.DCSIOLogic.LoadCleanLuasDCS() in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\DataStructures\DCS\DCSIOLogic.cs:line 258
[22/09/23 17:21:28]: 	Load default DCS: E:\Utilities\JoyPro\CleanProfile\DCS\A-10C.cf
[22/09/23 17:21:28]: 	Load default DCS: E:\Utilities\JoyPro\CleanProfile\DCS\A-10C_easy.cf
[22/09/23 17:21:28]: 	System.FormatException: Input string was not in a correct format.
[.... same data for some other profiles, like all "easy" profiles and also P47 and Mirage F1 at least ...]
[22/09/23 17:21:28]: 	Clean Joy Data loaded

```

But after this, JoyPro is loaded and I enter in its GUI where everything is empty (which is normal, as usual).

Then, I go into "import bindings", I select all, I click on "Import", and I get a crash pretty quickly.
I see two files created :
"<timestamp>.UnhandledException" which contains :
```System.UnhandledExceptionEventArgs
#
System.FormatException: Input string was not in a correct format.
   at System.Number.ParseDouble(String value, NumberStyles options, NumberFormatInfo numfmt)
   at System.Convert.ToDouble(String value, IFormatProvider provider)
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 112
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 105
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 105
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 105
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 105
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 105
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 105
   at JoyPro.DCSLuaInput.AnalyzeRawLuaInput(String content, DCSExportPlane refMod) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\DataStructures\DCS\DCSLuaInput.cs:line 333
   at JoyPro.DCSIOLogic.LoadLocalBinds(String localPath, List`1 planes, Boolean fillWithDefaults, String ending, Dictionary`2 resultsDict) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\DataStructures\DCS\DCSIOLogic.cs:line 774
   at JoyPro.DCSIOLogic.BindsFromLocal(List`1 sticks, List`1 planes, Boolean loadDefaults, Boolean inv, Boolean slid, Boolean curv, Boolean dz, Boolean sx, Boolean sy) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\DataStructures\DCS\DCSIOLogic.cs:line 884
   at JoyPro.ImportWindow.Import(Object sender, EventArgs e) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\Windows\ImportWindow.xaml.cs:line 185
   at System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   at System.Windows.UIElement.RaiseEventImpl(DependencyObject sender, RoutedEventArgs args)
   at System.Windows.Controls.Primitives.ButtonBase.OnClick()
   at System.Windows.Controls.Button.OnClick()
   at System.Windows.Controls.Primitives.ButtonBase.OnMouseLeftButtonUp(MouseButtonEventArgs e)
   at System.Windows.RoutedEventArgs.InvokeHandler(Delegate handler, Object target)
   at System.Windows.RoutedEventHandlerInfo.InvokeHandler(Object target, RoutedEventArgs routedEventArgs)
   at System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   at System.Windows.UIElement.ReRaiseEventAs(DependencyObject sender, RoutedEventArgs args, RoutedEvent newEvent)
   at System.Windows.UIElement.OnMouseUpThunk(Object sender, MouseButtonEventArgs e)
   at System.Windows.RoutedEventArgs.InvokeHandler(Delegate handler, Object target)
   at System.Windows.RoutedEventHandlerInfo.InvokeHandler(Object target, RoutedEventArgs routedEventArgs)
   at System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   at System.Windows.UIElement.RaiseEventImpl(DependencyObject sender, RoutedEventArgs args)
   at System.Windows.UIElement.RaiseTrustedEvent(RoutedEventArgs args)
   at System.Windows.Input.InputManager.ProcessStagingArea()
   at System.Windows.Input.InputManager.ProcessInput(InputEventArgs input)
   at System.Windows.Input.InputProviderSite.ReportInput(InputReport inputReport)
   at System.Windows.Interop.HwndMouseInputProvider.ReportInput(IntPtr hwnd, InputMode mode, Int32 timestamp, RawMouseActions actions, Int32 x, Int32 y, Int32 wheel)
   at System.Windows.Interop.HwndMouseInputProvider.FilterMessage(IntPtr hwnd, WindowMessage msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at System.Windows.Interop.HwndSource.InputFilterMessage(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndWrapper.WndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(Object o)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
   at System.Windows.Threading.Dispatcher.LegacyInvokeImpl(DispatcherPriority priority, TimeSpan timeout, Delegate method, Object args, Int32 numArgs)
   at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam)
   at MS.Win32.UnsafeNativeMethods.DispatchMessage(MSG& msg)
   at System.Windows.Threading.Dispatcher.PushFrameImpl(DispatcherFrame frame)
   at System.Windows.Application.RunDispatcher(Object ignore)
   at System.Windows.Application.RunInternal(Window window)
   at JoyPro.App.Main()
```

and "<timestamp>.DispUnhandledException" which contains :
```
System.Windows.Threading.DispatcherUnhandledExceptionEventArgs
#
Input string was not in a correct format.
#
   at System.Number.ParseDouble(String value, NumberStyles options, NumberFormatInfo numfmt)
   at System.Convert.ToDouble(String value, IFormatProvider provider)
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 112
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 105
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 105
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 105
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 105
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 105
   at JoyPro.LUADataRead.CreateAttributeDictFromLua(String cont) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\General\LUADataRead.cs:line 105
   at JoyPro.DCSLuaInput.AnalyzeRawLuaInput(String content, DCSExportPlane refMod) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\DataStructures\DCS\DCSLuaInput.cs:line 333
   at JoyPro.DCSIOLogic.LoadLocalBinds(String localPath, List`1 planes, Boolean fillWithDefaults, String ending, Dictionary`2 resultsDict) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\DataStructures\DCS\DCSIOLogic.cs:line 774
   at JoyPro.DCSIOLogic.BindsFromLocal(List`1 sticks, List`1 planes, Boolean loadDefaults, Boolean inv, Boolean slid, Boolean curv, Boolean dz, Boolean sx, Boolean sy) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\DataStructures\DCS\DCSIOLogic.cs:line 884
   at JoyPro.ImportWindow.Import(Object sender, EventArgs e) in D:\Dropbox\Programmierung\c#\JoyPro\JoystickProfiler\JoyPro\JoyPro\Windows\ImportWindow.xaml.cs:line 185
   at System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   at System.Windows.UIElement.RaiseEventImpl(DependencyObject sender, RoutedEventArgs args)
   at System.Windows.Controls.Primitives.ButtonBase.OnClick()
   at System.Windows.Controls.Button.OnClick()
   at System.Windows.Controls.Primitives.ButtonBase.OnMouseLeftButtonUp(MouseButtonEventArgs e)
   at System.Windows.RoutedEventArgs.InvokeHandler(Delegate handler, Object target)
   at System.Windows.RoutedEventHandlerInfo.InvokeHandler(Object target, RoutedEventArgs routedEventArgs)
   at System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   at System.Windows.UIElement.ReRaiseEventAs(DependencyObject sender, RoutedEventArgs args, RoutedEvent newEvent)
   at System.Windows.UIElement.OnMouseUpThunk(Object sender, MouseButtonEventArgs e)
   at System.Windows.RoutedEventArgs.InvokeHandler(Delegate handler, Object target)
   at System.Windows.RoutedEventHandlerInfo.InvokeHandler(Object target, RoutedEventArgs routedEventArgs)
   at System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   at System.Windows.UIElement.RaiseEventImpl(DependencyObject sender, RoutedEventArgs args)
   at System.Windows.UIElement.RaiseTrustedEvent(RoutedEventArgs args)
   at System.Windows.Input.InputManager.ProcessStagingArea()
   at System.Windows.Input.InputManager.ProcessInput(InputEventArgs input)
   at System.Windows.Input.InputProviderSite.ReportInput(InputReport inputReport)
   at System.Windows.Interop.HwndMouseInputProvider.ReportInput(IntPtr hwnd, InputMode mode, Int32 timestamp, RawMouseActions actions, Int32 x, Int32 y, Int32 wheel)
   at System.Windows.Interop.HwndMouseInputProvider.FilterMessage(IntPtr hwnd, WindowMessage msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at System.Windows.Interop.HwndSource.InputFilterMessage(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndWrapper.WndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(Object o)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
#
System.Collections.ListDictionaryInternal
#
```

I'm pretty sure it's a dumb thing but I spent the whole day to try things with no luck. I even returned to an older version of JoyPro, same behavior...

Do you have an idea on the problem ?

**Comment by Holdi601** [comment]
Can you upload your 2 cf files which cause the issue? also given as far as i can see its only the _easy planes and not the full fidelity version of the planes, you could just delete those files given the other files do not cause issues either. Waiting for your message then ill try and fix it

**Comment by Chaussette23** [comment]
Hello,

I continue to dig, and I may have found some other reasons, regarding the last time I used JoyPro (where it worked).
- I switched to Windows 11
- I moved DCS installation to a new drive. Joy pro is still in E:, SavedGames is still in F: but now DCS installation is in G:. That being said, DCS installation is not filled in settings, as I never used the DB thing
- I just bought new modules : P51, Apache, Hind (hence my usage of JoyPro)

I checked vcredist, they are all installed (and I think in this case, JoyPro wouldn't launch at all ?)

About failing cf files, here is **a third** of the list (too long...) :
- A-10C II_easy.cf
- A-10C_easy.cf
- Bf-109K-4_easy.cf
- F-16C_50_easy.cf
- F-5E-3_easy.cf
- F-86F_easy.cf
- FA-18C_hornet_easy.cf
- FW-190A8_easy.cf
- FW-190D9_easy.cf
- ka-50_easy.cf
- mi-8mt_easy.cf
- MiG-15bis_easy.cf
- Mirage-F1BE.cf
- Mirage-F1CE.cf
- Mirage-F1EE.cf
- Mirage-F1M-EE.cf
- MosquitoFBMkVI_easy.cf
- P-47D-30.cf
- P-47D-30_easy.cf
- P-51D-30-NA.cf
- P-51D-30-NA_easy.cf
- P-51D.cf
- SA342.cf
- SpitfireLFMkIX.cf
- SpitfireLFMkIXCW.cf
- SpitfireLFMkIXCW_easy.cf
- ...

I opened these cf files in Notepad++, I also compared them with working ones, everything looks fine, they are not truncated, they all finish with "return diff"+a new line.

Please find attached a working and a failing file, as well as the log files
[cf files example.zip](https://github.com/Holdi601/JoystickProfiler/files/12705907/cf.files.example.zip)
[logs.zip](https://github.com/Holdi601/JoystickProfiler/files/12705908/logs.zip)

In any case, thank you for having a look at this !

**Comment by Chaussette23** [comment]
Oh, I found way more precise things !

I cleared my input folder: import bindings doesn't crash (but doesn't import anything, as expected)
I restored only one input plane: A10C. I get a crash with only that one.

So I removed everything from that folder and tried diff.lua files one by one.

Only one diff.lua file makes JoyPro crashing: it's the one from my throttle !

I attached it, that should be easier for you to reproduce and debug:
[Input.zip](https://github.com/Holdi601/JoystickProfiler/files/12706004/Input.zip)

**Comment by Chaussette23** [comment]
I found the problem !

I made a minimalist diff.lua file. The crash occur when there is a curvature value with a decimal part on an axis curvature.

["curvature"] = { [1] = 0,} WORKS
["curvature"] = { [1] = 1,} WORKS
["curvature"] = { [1] = 0.45,} FAIL (value I've set)

The last time I used JoyPro (in working condition) was in Windows 10 in French (with standard French region settings). However, I remember that I upgraded to Windows 11, but I've set Windows to english language with French number formats.

Here is my localisation settings:
![image](https://github.com/Holdi601/JoystickProfiler/assets/119120762/3362981c-51f5-4433-9943-87e76ebfb941)

I just reset my localization parameters (reset button), and now JoyPro is working fine, even with the minimalist diff.lua file.

So to reproduce the problem, you "just" have to replace the decimal symbol from point '.' to comma ','.

On my side, I will keep the point instead of the comma because I don't care, so I close the ticket !
